### PR TITLE
doc: document public api

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

**Review by fl0rek — COMMENTED** [review]
nicely done, few nitpics :)

File: types/src/blob.rs [hunk 1]
```diff
@@ -43,6 +86,29 @@ impl Blob {
...
+    ///
+    /// If validation fails, this function will return an error with a reason of failure.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use celestia_types::Blob;
+    /// # use celestia_types::nmt::Namespace;
+    /// #
+    /// # let namespace = Namespace::new_v0(&[1, 2, 3, 4, 5]).expect("Invalid namespace");
```
[the remark this answers is not in the document]
Comment (zvolin) [reply types/src/blob.rs:101]: it hides this code from the example in docs

File: types/src/consts.rs [hunk 2]
```diff
@@ -1,88 +1,105 @@
...
+        /// The number of bytes reserved for the share metadata.
+        ///
+        /// The info byte contains the share version and a sequence start idicator.
         pub const SHARE_INFO_BYTES: usize = 1;
 
-        /// the number of bytes reserved for the sequence length
-        /// that is present in the first share of a sequence.
+        /// The number of bytes reserved for the sequence length in a share.
+        ///
+        /// That is present in the first share of a sequence.
```
[the remark this answers is not in the document]
Comment (zvolin) [reply types/src/consts.rs:54]: hmmm, why not
```suggestion
        /// The number of bytes reserved for the sequence length in a share.
        /// It is present only in the first share of a sequence.
```

File: types/src/hash.rs [hunk 3]
```diff
@@ -1,6 +1,15 @@
+//! Celestia hash related types and traits.
+
+/// The hash type used commonly in the Celestia.
 pub type Hash = tendermint::hash::Hash;
 
+/// A trait extending the [`Hash`] functionalities.
+///
+/// [`Hash`]: crate::hash::Hash
 pub trait HashExt {
+    /// Get the default `SHA256` hash.
```
[the remark this answers is not in the document]
Comment (zvolin) [reply types/src/hash.rs:10]: wdyt about that?

```suggestion
    /// Get the `SHA256` hash of an empty input.
```

File: types/src/nmt.rs [hunk 4]
```diff
@@ -90,6 +249,16 @@ impl Namespace {
         Namespace(nmt_rs::NamespaceId(bytes))
     }
 
+    /// Create a new [`Namespace`] version `0` with given id.
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/nmt.rs:252]: I'll apply the trivial ones, feel free to roll back

File: node/src/p2p.rs [hunk 5]
```diff
@@ -1,3 +1,15 @@
+//! Component responsible for the messeging and interacting on peer to peer layer with other nodes.
+//!
+//! It is a high level integration of various p2p protocols used by Celestia nodes.
+//! Currently supporting:
```
[the remark this answers is not in the document]
Comment (zvolin) [reply node/src/p2p.rs:4]: our roadmap is rather too noisy, I don't think protocol list will change often so maybe we can try to keep it up to date, or remove it completely

File: types/src/share.rs [hunk 6]
```diff
@@ -18,33 +18,66 @@ mod info_byte;
...
+///
+/// All shares have the fixed size of 512 bytes and the following structure:
+///
+/// ```text
+/// | Namespace | InfoByte | (optional) sequence length | data |
+/// ```
+///
+/// The `sequence length` field indicates the byte length of the data split into shares.
+/// If the data split into shares cannot fit into a single share, then each following
+/// share shouldn't have this field set.
```
[the remark this answers is not in the document]
Comment (zvolin) [reply types/src/share.rs:52]: Oh I remember how I sturggled there :D

File: types/src/nmt.rs [hunk 7]
```diff
@@ -18,34 +37,128 @@ pub use self::namespaced_hash::{
...
+    /// Primary reserved [`Namespace`] for the compact [`Share`]s with [`MsgPayForBlobs`] transactions.
+    ///
+    /// [`Share`]: crate::share::Share
+    /// [`MsgPayForBlobs`]: celestia_proto::celestia::blob::v1::MsgPayForBlobs
     pub const PAY_FOR_BLOB: Namespace = Namespace::const_v0([0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
+
+    /// Primary reserved [`Namespace`] for the [`Share`]s used for padding.
+    ///
+    /// It is used for [`Share`]s inserted after all other shares from primary reserved namespace
+    /// to align user-defined namespaces correctly in [`ExtendedDataSquare`].
```
Comment (fl0rek) [inline types/src/nmt.rs:109]:
```suggestion
    /// [Share]s with this namespace are inserted after other shares from primary reserved namespace
    /// so that user-defined namespaces are correctly aligned in [`ExtendedDataSquare`]
```
idk if it's me, but I find original hard to parse - feel free to include or not

File: types/src/nmt.rs [hunk 8]
```diff
@@ -62,6 +191,36 @@ impl Namespace {
...
+    /// The `id` must be either:
+    ///  - a 28 byte slice specifying full id
+    ///  - a 10 or less byte slice specifying user-defined suffix
+    ///
+    /// [`Namespace`]s in version 0 must have id's prefixed with 18 `0x00` bytes.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the provided id is of incorrect length
+    /// or if the `id` was of 28 bytes and have incorrect `0x00` prefix.
```
Comment (fl0rek) [inline types/src/nmt.rs:205]:
```suggestion
    /// This function will return an error if the provided id has incorrect length
    /// or if the `id` has 28 bytes and have doesn't have mandatory 18x`0x00` bytes prefix
```

File: types/src/nmt.rs [hunk 9]
```diff
@@ -108,12 +277,42 @@ impl Namespace {
...
     pub const fn const_v255(id: u8) -> Self {
         let mut bytes = [255u8; NS_SIZE];
         bytes[NS_ID_SIZE] = id;
         Namespace(nmt_rs::NamespaceId(bytes))
     }
 
+    /// Create a new [`Namespace`] version `255` with a given id.
+    ///
+    /// [`Namespace`]s with version `255` must have ids prefixed with 27 `0xff` bytes followed by a
+    /// single byte with actual id.
```
Comment (fl0rek) [inline types/src/nmt.rs:300]:
```suggestion
    /// single byte with an actual id.
```

File: types/src/nmt.rs [hunk 10]
```diff
@@ -108,12 +277,42 @@ impl Namespace {
...
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the provided id is of incorrect length
+    /// or if the `id` prefix is incorrect.
```
Comment (fl0rek) [inline types/src/nmt.rs:305]:
```suggestion
    /// This function will return an error, if the provided id has incorrect length
    /// or if the `id` prefix is incorrect.
```

File: types/src/nmt.rs [hunk 11]
```diff
@@ -129,18 +328,22 @@ impl Namespace {
         }
     }
 
+    /// Converts the [`Namespace`] to the byte slice.
```
Comment (fl0rek) [inline types/src/nmt.rs:331]:
```suggestion
    /// Converts the [`Namespace`] to a byte slice.
```

File: types/src/nmt/namespace_proof.rs [hunk 12]
```diff
@@ -10,22 +10,83 @@ use crate::{Error, Result};
...
+    /// [`AbsenceProof`]: NmtNamespaceProof::PresenceProof
+    /// [`NamespacedHash`]: crate::nmt::NamespacedHash
     pub fn leaf(&self) -> Option<&NamespacedHash> {
         match &self.0 {
             NmtNamespaceProof::AbsenceProof { leaf, .. } => leaf.as_ref(),
             _ => None,
         }
     }
 
+    /// Returns true if the proof will ignore all the leaves inserted with
```
Comment (fl0rek) [inline types/src/nmt/namespace_proof.rs:86]:
```suggestion
    /// Returns true if the proof ignores all the leaves inserted with
```

File: types/src/nmt/namespaced_hash.rs [hunk 13]
```diff
@@ -3,16 +3,32 @@ use crate::{Error, Result};
...
 
+/// An extention trait for the [`NamespacedHash`] to perform additional actions.
 pub trait NamespacedHashExt {
+    /// Get the hash of the root of an empty [`Nmt`].
+    ///
+    /// [`Nmt`]: crate::nmt::Nmt
     fn empty_root() -> NamespacedHash;
+    /// Try to decode [`NamespacedHash`] from the raw bytes.
     fn from_raw(bytes: &[u8]) -> Result<NamespacedHash>;
+    /// Encode [`NamespacedHash`] to the [`Vec`].
```
Comment (fl0rek) [inline types/src/nmt/namespaced_hash.rs:22]:
```suggestion
    /// Encode [`NamespacedHash`] into [`Vec`].
```

File: types/src/nmt/namespaced_hash.rs [hunk 14]
```diff
@@ -3,16 +3,32 @@ use crate::{Error, Result};
...
     fn to_vec(&self) -> Vec<u8>;
+    /// Encode [`NamespacedHash`] to the [`array`].
```
Comment (fl0rek) [inline types/src/nmt/namespaced_hash.rs:24]:
```suggestion
    /// Encode [`NamespacedHash`] into [`array`].
```

File: types/src/rsmt2d.rs [hunk 15]
```diff
@@ -1,8 +1,109 @@
...
+/// The data matrix in Celestia blocks extended with parity data.
+///
+/// It is created by a fixed size chunks of data, called [`Share`]s.
+/// Each share is a cell of the [`ExtendedDataSquare`].
+///
+/// # Structure
+///
+/// The [`ExtendedDataSquare`] consists of four quadrants. The first
+/// quadrant (upper-left) is the original data submitted to the network,
+/// referred to as `OriginalDataSquare`. The rest three quadrants are
```
Comment (fl0rek) [inline types/src/rsmt2d.rs:12]:
```suggestion
/// referred to as `OriginalDataSquare`. The other three quadrants are
```

File: types/src/sync.rs [hunk 16]
```diff
@@ -2,15 +2,25 @@ use serde::{Deserialize, Serialize};
 use tendermint::hash::Hash;
 use tendermint::time::Time;
 
+/// A state of the blockchain synchronization.
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct SyncState {
+    /// The ID of current syncing process.
```
Comment (fl0rek) [inline types/src/sync.rs:8]:
```suggestion
    /// The ID of the current syncing process.
```

File: types/src/sync.rs [hunk 17]
```diff
@@ -2,15 +2,25 @@ use serde::{Deserialize, Serialize};
...
     pub to_height: u64,
+    /// The first hash to be synced.
     pub from_hash: Hash,
+    /// The last hash to be synced.
     pub to_hash: Hash,
+    /// The time when syncing began.
     pub start: Time,
+    /// The time when syncing ended.
     pub end: Time,
+    /// The error that could happen during synchronization.
```
Comment (fl0rek) [inline types/src/sync.rs:24]:
```suggestion
    /// Any error during synchronisation, if it occured
```
Current wording makes it sound like we were anticipating an error 😅

**Review by fl0rek — APPROVED** [review]
Nice work, our docs look much more proffesional now 🎸 
Minor suggestions, mostly definite/indefinite articles :D

File: node/src/network.rs [hunk 18]
```diff
@@ -1,19 +1,27 @@
+//! Primitives and constants related to the networks supported by Celestia nodes.
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply node/src/network.rs:1]: actually nvm, I think it's good

File: types/src/nmt.rs (hunk 7 above)
Comment (zvolin) [reply types/src/nmt.rs:109]: sure, just corrected backticks

File: types/src/state/tx.rs [hunk 19]
```diff
@@ -1,14 +1,17 @@
 use celestia_proto::cosmos::base::abci::v1beta1::TxResponse as RawTxResponse;
 use serde::{Deserialize, Serialize};
 
-/// Raw TX data
+/// Raw transaction data.
 ///
-/// NOTE: Tx has no types at this level
+/// # Note
+///
+/// Tx has no types at this level
```
Comment (oblique) [inline types/src/state/tx.rs:8]:
```suggestion
/// Transaction has no types at this level
```

**Review by oblique — APPROVED** [review]
Good work!! I have some minor comments.

File: blockstore/src/lib.rs [hunk 20]
```diff
@@ -1,3 +1,4 @@
+#![doc = include_str!("../README.md")]
 #![cfg_attr(docs_rs, feature(async_fn_in_trait))]
```
[the remark this answers is not in the document]
Comment (zvolin) [reply blockstore/src/lib.rs:2]:
```suggestion
```

**Review by oblique — APPROVED** [review]
lgtm
